qiskit/algorithms/optimizers/sam.py

```diff
@@ -0,0 +1,144 @@
+            'gradient': OptimizerSupportLevel.ignored,
+            'bounds': OptimizerSupportLevel.ignored,
+            'initial_point': OptimizerSupportLevel.required
+        }
+
+    def minimize(self, objective_function: Callable[[np.ndarray], float], initial_point: np.ndarray,
+                 gradient_function: Callable[[np.ndarray], float]) -> Tuple[np.ndarray, float, int]:
+        """Run the minimization.
+
+        Args:
+            objective_function: A function handle to the objective function.
+            initial_point: The initial iteration point.
+            gradient_function: A function handle to the gradient of the objective function.
+
+        Returns:
+            A tuple of (optimal parameters, optimal value, number of iterations).
+        """
+
+        # Naive implementation
+        params = params_new = initial_point
+        for it in range(self._maxiter):
+            grad = gradient_function(params)
+            sam = params + self._rho * grad / np.linalg.norm(grad)  # euclid norm is optimal
+            grad_sam = gradient_function(sam)
+            params_new = params - self._eta * grad_sam / np.linalg.norm(grad_sam)
+
+            if np.linalg.norm(params - params_new) < self._tol:
+                return params_new, objective_function(params_new), it + 1
+            params = params_new
+
+        return params_new, objective_function(params_new), self._maxiter
+
+    def optimize(self, num_vars: int, objective_function: Callable[[np.ndarray], float],
+                 gradient_function: Optional[Callable[[np.ndarray], float]] = None,
+                 variable_bounds: Optional[List[Tuple[float, float]]] = None,
+                 initial_point: Optional[np.ndarray] = None
+                 ) -> Tuple[np.ndarray, float, int]:
+        """Perform optimization.
+
+        Args:
+            num_vars: Number of parameters to be optimized.
+            objective_function: Handle to a function that computes the objective function.
+            gradient_function: Handle to a function that computes the gradient of the objective
+                function.
+            variable_bounds: deprecated
+            initial_point: The initial point for the optimization.
+
+        Returns:
+            A tuple (point, value, nfev) where\n
+                point: is a 1D numpy.ndarray[float] containing the solution\n
+                value: is a float with the objective function value\n
+                nfev: is the number of objective function calls
+        """
+        super().optimize(num_vars, objective_function, gradient_function,
+                         variable_bounds, initial_point)
+        if initial_point is None:
+            initial_point = algorithm_globals.random.random(num_vars)
+        if gradient_function is None:
+            gradient_function = Optimizer.wrap_function(Optimizer.gradient_num_diff,
+                                                        (objective_function, self._eps))
```

Comment:
If you look at SLSQP you will see that this function can also take the max evals grouped. Since these scipy optimizers have their own finite diff as a default, there we only use this one when max evals grouped is not one. The goal there was to be able to group up more than one point as the gradient was computed to minimise overhead and potentially with Aer to allow more to be run in parallel. You will see that we can take advantage of this via algorithms such as VQE where we can accept a list of points on the object function not just a single point. It defaults to 1 for the single point that would be normally expected of course.

Reply:
Thanks @woodsp-ibm.
In SLSQP, I do not really understand that statement https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/algorithms/optimizers/slsqp.py#L74
because `_max_evals_grouped` can not be set in SLSQP, hence is default to 1 and `if` condition is always `false`

Reply:
sorry, there is setter in `Optimizer` class, you can disregard my last comment

Reply:
`_max_evals_grouped` enabled, I will test it in VQE when I am finished